clarity/src/vm/callables.rs

```diff
@@ -168,30 +172,65 @@ impl DefinedFunction {
         for arg in arg_iterator.drain(..) {
             let ((name, type_sig), value) = arg;
 
-            match (type_sig, value) {
+            // Arguments containing principal literals can be implicitly cast to traits
+            // to match parameter types.
+            // e.g. `(some .foo)` to `(optional <trait>`)
+            // and traits can be implicitly cast to sub-traits
+            // e.g. `<foo-and-bar>` to `<foo>`
+            let cast_value =
+                if *env.contract_context.get_clarity_version() >= ClarityVersion::Clarity2 {
```

Comment:
It might be worthwhile to wrap this block into a method so we can isolate the version gating.

Reply:
Given the documented Clarity1-only behavior below, might it make sense to keep all of the Clarity1-dependent logic in its own method (or top-level block in this method)?

Reply:
I'm not sure what would be an improvement here. Are you suggesting something like:

```rust
        for arg in arg_iterator.drain(..) {
            let ((name, type_sig), value) = arg;

            if *env.contract_context.get_clarity_version() >= ClarityVersion::Clarity2 {
                self.process_arg_v1(name, type_sig, value);
            } else {
                self.process_arg(name, type_sig, value);
            }

            if let Some(_) = context.variables.insert(name.clone(), cast_value) {
                return Err(CheckErrors::NameAlreadyUsed(name.to_string()).into());
            }
        }
```

Reply:
I'm suggesting that the act of calculating `cast_value` be in its own method.

Ultimately, I'm trying to minimize the number of gates that exist in the code so we can preserve the old Clarity1 code's original structure as best we can.  I'd like to do this so that it'll be easier to manually verify that the old Clarity1 code continues to run as expected in 2.1.

Reply:
Refactored in 49b103d64.